pkg/apply/solver/solver.go

```diff
@@ -66,105 +67,133 @@ func (t *TaskQueueSolver) BuildTaskQueue(ro resourceObjects,
-			objs := object.UnstructuredsToObjMetas(crdSplitRes.crds)
-			tasks = append(tasks, taskrunner.NewWaitTask(
-				objs,
-				taskrunner.AllCurrent,
-				1*time.Minute),
-				&task.ResetRESTMapperTask{
-					Mapper: t.Mapper,
-				})
+			tasks = t.appendWaitTask(tasks, crdSplitRes.crds, 1*time.Minute)
+			tasks = append(tasks, &task.ResetRESTMapperTask{
+				Mapper: t.Mapper,
+			})
 		}
 		remainingInfos = crdSplitRes.after
 	}
 
-	tasks = append(tasks,
-		&task.ApplyTask{
-			Objects:           remainingInfos,
-			CRDs:              crdSplitRes.crds,
-			ServerSideOptions: o.ServerSideOptions,
-			DryRunStrategy:    o.DryRunStrategy,
-			InfoHelper:        t.InfoHelper,
-			Factory:           t.Factory,
-			Mapper:            t.Mapper,
-		},
-		&task.SendEventTask{
-			Event: event.Event{
-				Type: event.ApplyType,
-				ApplyEvent: event.ApplyEvent{
-					Type: event.ApplyEventCompleted,
-				},
-			},
-		},
-	)
+	tasks = t.appendApplyTask(tasks, remainingInfos, crdSplitRes.crds, o)
+	tasks = t.appendSendApplyEventTask(tasks)
 
 	if !o.DryRunStrategy.ClientOrServerDryRun() && o.ReconcileTimeout != time.Duration(0) {
-		tasks = append(tasks,
-			taskrunner.NewWaitTask(
-				ro.IdsForApply(),
-				taskrunner.AllCurrent,
-				o.ReconcileTimeout),
-			&task.SendEventTask{
-				Event: event.Event{
-					Type: event.StatusType,
-					StatusEvent: event.StatusEvent{
-						Type: event.StatusEventCompleted,
-					},
-				},
-			},
-		)
+		tasks = t.appendWaitTask(tasks, ro.ObjsForApply(), o.ReconcileTimeout)
+		tasks = t.appendSendStatusEventTask(tasks)
 	}
 
 	if o.Prune {
-		tasks = append(tasks,
-			&task.PruneTask{
-				Objects:           ro.ObjsForApply(),
-				InventoryObject:   ro.Inventory(),
-				PruneOptions:      t.PruneOptions,
-				PropagationPolicy: o.PrunePropagationPolicy,
-				DryRunStrategy:    o.DryRunStrategy,
-			},
-			&task.SendEventTask{
-				Event: event.Event{
-					Type: event.PruneType,
-					PruneEvent: event.PruneEvent{
-						Type: event.PruneEventCompleted,
-					},
-				},
-			},
-		)
+		tasks = t.appendPruneTask(tasks, ro.ObjsForApply(), ro.Inventory(), o)
+		tasks = t.appendSendPruneEventTask(tasks)
 
 		if !o.DryRunStrategy.ClientOrServerDryRun() && o.PruneTimeout != time.Duration(0) {
-			tasks = append(tasks,
-				taskrunner.NewWaitTask(
-					ro.IdsForPrune(),
-					taskrunner.AllNotFound,
-					o.PruneTimeout),
-				&task.SendEventTask{
-					Event: event.Event{
-						Type: event.StatusType,
-						StatusEvent: event.StatusEvent{
-							Type: event.StatusEventCompleted,
-						},
-					},
-				},
-			)
+			tasks = t.appendWaitPruneTask(tasks, ro.ObjsForApply(), o.PruneTimeout)
+			tasks = t.appendSendStatusEventTask(tasks)
 		}
 	}
 
 	return tasksToQueue(tasks)
 }
 
+func (t *TaskQueueSolver) appendApplyTask(tasks []taskrunner.Task,
+	objects []*unstructured.Unstructured,
+	crds []*unstructured.Unstructured,
+	o Options) []taskrunner.Task {
+	if o.ContinueOnError {
```

Comment:
I think we should look at other options before we implement this pattern. This can end up with a large number of tasks...

Reply:
I considered some other option, but they are not better than the proposed change. According to Sean, when kubectl library >= 1.18, `ContinueOnError` is the default behavior. If we adopt that and create an Apply task with multiple objects, `ContinueOnError` is supported by `AbortOnError` is not supported. If we create a bunch of Apply tasks, each applies one object, then we can support both `ContinueOnError` and `AbortOnError`. It becomes an option the library caller can choose.

Are you concerned about the performance of a large number of tasks? Where could be the bottleneck? If it's in creating the kubectl ApplyOption, we can optimize it to avoid recreating it later.

Reply:
Updated this to keep creating one apply task and avoid creating `ApplyOptions` multiple times. The difference will be within the ApplyTask. When `ContinueOnError` is true, the apply tasks calls kubectl `ApplyOptions.Run` multiple times.